src/partition/repart.c

```diff
@@ -3439,7 +3440,58 @@ static int partition_populate_filesystem(Partition *p, const char *node) {
         return 0;
 }
 
+static int make_copy_files_denylist(Context *context, const Partition *p, Set **ret) {
+        _cleanup_set_free_ Set *denylist = NULL;
+        int r;
+
+        assert(context);
+        assert(ret);
+
+        /* When populating root partitions, let's make sure we're not duplicating data from other partitions
+         * by adding those partition's canonical mount points to the deny list we pass to copy_tree(). */
+
+        const char *mountpoint = gpt_partition_type_mount_point(p->type_uuid);
+        if (!mountpoint)
+                return 0;
+
+        LIST_FOREACH(partitions, q, context->partitions) {
+                _cleanup_free_ char *d = NULL;
+                struct stat st;
+
+                if (p == q)
+                        continue;
+
+                const char *source = gpt_partition_type_mount_point(q->type_uuid);
+                if (!source)
+                        continue;
+
+                /* Make sure we only exclude nested partition mount points. */
+                if (isempty(path_startswith(source, mountpoint)))
+                        continue;
+
+                r = chase_symlinks_and_stat(source, arg_root, CHASE_PREFIX_ROOT|CHASE_WARN, NULL, &st, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to stat source file '%s%s': %m",
+                                               strempty(arg_root), source);
+
+                if (set_contains(denylist, &st))
+                        continue;
+
+                d = memdup(&st, sizeof(st));
+                if (!d)
+                        return log_oom();
+                if (set_ensure_put(&denylist, &inode_hash_ops, d) < 0)
+                        return log_oom();
+
+                TAKE_PTR(d);
+        }
+
+        *ret = TAKE_PTR(denylist);
+        return 0;
+}
```

Comment:
hmm, I am not sure I follow here. So here's what I had in mind:

Taking inspiration from `context_open_copy_block_paths()` add a function `context_open_copy_files_paths()`, that iterates through all partitions, and opens the source inodes to copy from, writing an fd for each top-level inode fd into a new field `p->copy_files_fd`. At the same time fstat() the file, and build up the denylist. When this new function `context_open_copy_files_paths()` is complete we thus have opened/pinned all top-level source inodes, and built a complete denylist based on their inode info. 

Then in a later step, actaully iterate throught the partitions again, now execute the copy routine, each time passing in the already open inode fd, plus the complete deny list.

I thin this would be relatively pretty because it mirrors the copy_blocks logic, and we have a global denylist for all operations.

Does this make sense?

Reply:
That seems to result in a overly restrictive denylist no?

Let's say I have the following in one partition definition:

```
CopyFiles=/etc/fstab:/home/<user>/fstab-copy
```

Just because we copy /etc/fstab to some location in the home directory doesn't mean that we don't want to copy it to /etc/fstab anymore.

I would very much prefer to have a minimal denylist per partition rather than a global one. We generally don't want to prevent users copying files to wherever they want, but only want to intervene when we're 99% sure the behavior doesn't make sense, and to me in this case that's only when we start duplicating full partitions into other partitions, so I'd like to just prevent that from happening but nothing else, at least not without explicit opt-in from the user.

Reply:
Had another look and I figure we can make the denylist global instead of per partition, that just means that we never copy any nested partitions into another partition, in practice, the behavior should be exactly the same.

I would prefer to stick to just excluding nested partition mount points instead of everything from CopyFiles= by default. If we need more, let's make it opt-in via an option.